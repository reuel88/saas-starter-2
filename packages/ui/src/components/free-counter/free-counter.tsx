"use client";

import { Zap } from "lucide-react";
import { FC, useEffect, useState } from "react";
import { Button } from "../button";
import { Progress } from "../ui/progress";

type FreeCounterProps = {
  count: number;
  max: number;
  onClick: () => void;
};

export const FreeCounter: FC<FreeCounterProps> = ({
  count,
  max,
  onClick,
  ...rest
}) => {
  const [mounted, setMounted] = useState(false);

  useEffect(() => {
    setMounted(true);
  }, []);

  if (!mounted) return null;

  return (
    <div className="flex items-center gap-4" {...rest}>
      <div className="space-y-2 whitespace-nowrap text-sm text-secondary-foreground">
        <p id="progress-label">
          {count} / {max} Free Generations
        </p>
        <Progress
          aria-labelledby="progress-label"
          className="h-3"
          value={(count / max) * 100}
        />
      </div>
      <Button onClick={onClick} variant="premium" className="w-full">
        Upgrade
        <Zap className="ml-2 h-4 w-4 fill-white" />
      </Button>
    </div>
  );
};
