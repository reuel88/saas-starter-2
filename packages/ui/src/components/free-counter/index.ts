export { FreeCounter } from "./free-counter";
