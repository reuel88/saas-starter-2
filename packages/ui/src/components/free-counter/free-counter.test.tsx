import { render, screen } from "@testing-library/react";
import { FreeCounter } from "./free-counter";
import { vi } from "vitest";
import userEvent from "@testing-library/user-event";

describe("FreeCounter", () => {
  it("should render correctly", () => {
    // Arrange
    const onClickMock = vi.fn();

    // Act
    const { baseElement } = render(
      <FreeCounter
        count={0}
        max={10}
        onClick={onClickMock}
        data-testid="free-counter"
      />,
    );
    const divElement = screen.getByTestId("free-counter");

    // Assert
    expect(divElement).toHaveClass("flex items-center gap-4");
    expect(baseElement).toMatchSnapshot();
  });

  it("should call onClick when button is clicked", async () => {
    // Arrange
    const onClickMock = vi.fn();

    // Act
    render(
      <FreeCounter
        count={0}
        max={10}
        onClick={onClickMock}
        data-testid="free-counter"
      />,
    );
    const buttonElement = screen.getByRole("button");
    await userEvent.click(buttonElement);

    // Assert
    expect(onClickMock).toHaveBeenCalled();
  });
});
