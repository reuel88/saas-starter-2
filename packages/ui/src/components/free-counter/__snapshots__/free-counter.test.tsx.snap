// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FreeCounter > should render correctly 1`] = `
<body>
  <div>
    <div
      class="flex items-center gap-4"
      data-testid="free-counter"
    >
      <div
        class="space-y-2 whitespace-nowrap text-sm text-secondary-foreground"
      >
        <p
          id="progress-label"
        >
          0
           / 
          10
           Free Generations
        </p>
        <div
          aria-labelledby="progress-label"
          aria-valuemax="100"
          aria-valuemin="0"
          class="relative w-full overflow-hidden rounded-full bg-secondary h-3"
          data-max="100"
          data-state="indeterminate"
          role="progressbar"
        >
          <div
            class="h-full w-full flex-1 bg-primary transition-all"
            data-max="100"
            data-state="indeterminate"
            style="transform: translateX(-100%);"
          />
        </div>
      </div>
      <button
        class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-gradient-to-r from-indigo-500 via-purple-500 to-pink-500 text-white border-0 h-10 px-4 py-2 w-full"
      >
        Upgrade
        <svg
          class="lucide lucide-zap ml-2 h-4 w-4 fill-white"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <polygon
            points="13 2 3 14 12 14 11 22 21 10 12 10 13 2"
          />
        </svg>
      </button>
    </div>
  </div>
</body>
`;
